src/hash.cr

```diff
@@ -460,20 +470,25 @@ class Hash(K, V)
     while true
       entry_index = get_index(index)
 
-      # If we find an empty index slot, there's no such key
-      if entry_index == -1
+      # If we find an empty index slot, there are no more keys to search
+      if entry_index == EMPTY_INDEX
         return nil
       end
 
-      # We found a non-empty slot, let's see if the key we have matches
-      entry = get_entry(entry_index)
-      if entry_matches?(entry, hash, key)
-        delete_entry_and_update_counts(entry_index)
-        return entry
-      else
-        # If it doesn't, check the next index...
-        index = next_index(index)
+      # Skip over deleted index slots
+      unless entry_index == DELETED_INDEX
+        # We found a non-empty slot, let's see if the key we have matches
+        entry = get_entry(entry_index)
+        if entry_matches?(entry, hash, key)
+          # Mark this index slot as deleted
+          delete_index(index)
+          delete_entry_and_update_counts(entry_index)
+          return entry
+        end
       end
+
+      # Nope, move on to the next slot
+      index = next_index(index)
```

Comment:
**polish:** We could consider combining these two branches in a `case` expression. Completely optional, of course.
Same for `find_entry_with_index`.